Subject: Linux-RT 2.6.24-rc2-rt1
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From dbahi@novell.com Wed Sep 26 14:50:03 2007
Return-path: <dbahi@novell.com>
X-Spam-Checker-Version: SpamAssassin 3.2.1 (2007-05-02) on
	gandalf.stny.rr.com
X-Spam-Level: 
X-Spam-Status: No, score=-2.5 required=5.0 tests=AWL,BAYES_00,RDNS_NONE
	autolearn=no version=3.2.1
Envelope-to: rostedt@localhost
Delivery-date: Wed, 26 Sep 2007 14:50:03 -0400
Received: from localhost ([127.0.0.1] helo=localhost.localdomain
	ident=rostedt) by gandalf.stny.rr.com with esmtp (Exim 4.67) (envelope-from
	<dbahi@novell.com>) id 1IabxS-00031Z-Va for rostedt@localhost; Wed, 26 Sep
	2007 14:50:03 -0400
X-Real-To: <rostedt@goodmis.org>
X-Trusted: false
Received: from webmaillogin.com [216.40.35.34] by localhost.localdomain
	with IMAP (fetchmail-6.3.8) for <rostedt@localhost> (single-drop); Wed, 26
	Sep 2007 14:50:02 -0400 (EDT)
Received: from emd2-ims8.emaildefenseservice.com ([64.97.158.115] verified)
	by fr3.webmaillogin.com (CommuniGate Pro SMTP 5.0.9) with SMTP id 366482337
	for rostedt@goodmis.org; Wed, 26 Sep 2007 14:45:32 -0400
Received: from emd2-imf3.emaildefenseservice.com
	(emd2-ims.emaildefenseservice.com [64.97.158.10]) by
	emd2-ims8.emaildefenseservice.com (EMD Outbound ESMTP Server) with ESMTP id
	E9A4353CC09 for <rostedt@goodmis.org>; Wed, 26 Sep 2007 14:45:31 -0400 (EDT)
X-SpamScore: Exempt
Received: from sinclair.provo.novell.com (sinclair.provo.novell.com
	[137.65.248.137]) by emd2-imf3.emaildefenseservice.com (Postfix) with ESMTP
	for <rostedt@goodmis.org>; Wed, 26 Sep 2007 14:45:31 -0400 (EDT)
Received: from INET-PRV-MTA by sinclair.provo.novell.com with
	Novell_GroupWise; Wed, 26 Sep 2007 12:45:25 -0600
Message-Id: <46FA545D0200006C0001930D@sinclair.provo.novell.com>
X-Mailer: Novell GroupWise Internet Agent 7.0.2 HP
Date: Wed, 26 Sep 2007 12:45:17 -0600
From: "David Bahi" <dbahi@novell.com>
To: <linux-kernel@vger.kernel.org>,<linux-rt-users@vger.kernel.org>
Cc: <rostedt@goodmis.org>,<tglx@linutronix.de>, "Gregory Haskins" <GHaskins@novell.com>
Subject: [PATCH 1/1] nmi_watchdog: x86_64 count timer and hpet like i386
References: <46FA4AE7020000FC00012412@sinclair.provo.novell.com>
	 <46FA545D0200006C0001930D@sinclair.provo.novell.com>
In-Reply-To: <46FA545D0200006C0001930D@sinclair.provo.novell.com>
Mime-Version: 1.0
Content-Type: multipart/mixed; boundary="=__PartD1F660AD.0__="
Status: RO
X-Status: 
X-Keywords:                 
X-UID: 193
X-Evolution-Source: mbox:///var/mail/rostedt


--=__PartD1F660AD.0__=
Content-Type: text/plain; charset=US-ASCII
Content-Disposition: inline
Content-Transfer-Encoding: 8bit

This modifies nmi_watchdog_tick behavior for x86_64 arch to 
consider both timer and pit/hpet IRQs just as the i386 arch does.

Without this fix a kernel crash occurs very early in the boot 
process if nmi_watchdog is on.

Signed-off-by: David Bahi <dbahi@novell.com>

---
 arch/x86_64/kernel/nmi.c |    8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

Index: linux-2.6.22.8-rt9_1267/arch/x86_64/kernel/nmi.c
===================================================================
--- linux-2.6.22.8-rt9_1267.orig/arch/x86/kernel/nmi_64.c
+++ linux-2.6.22.8-rt9_1267/arch/x86/kernel/nmi_64.c
@@ -369,8 +369,6 @@ int notrace __kprobes nmi_watchdog_tick(
                touched = 1;
        }
 
-       sum = read_pda(apic_timer_irqs);
-
        if (__get_cpu_var(nmi_touch)) {
                __get_cpu_var(nmi_touch) = 0;
                touched = 1;
@@ -386,6 +384,12 @@ int notrace __kprobes nmi_watchdog_tick(
                cpu_clear(cpu, backtrace_mask);
        }
 
+       /*
+        * Take the local apic timer and PIT/HPET into account. We don't
+        * know which one is active, when we have highres/dyntick on
+        */
+       sum = read_pda(apic_timer_irqs) + kstat_cpu(cpu).irqs[0];
+
 #ifdef CONFIG_X86_MCE
        /* Could check oops_in_progress here too, but it's safer
           not too */



--=__PartD1F660AD.0__=
Content-Type: application/pgp-signature; name="signature.asc"
Content-Transfer-Encoding: base64
Content-Disposition: attachment; filename="signature.asc"

LS0tLS1CRUdJTiBQR1AgU0lHTkFUVVJFLS0tLS0KVmVyc2lvbjogR251UEcgdjEuNC4yIChHTlUv
TGludXgpCgppRDhEQlFCRytxaTJIMjhKSFlhekxhUVJBbEJyQUtDSWxLOTczRVNJNmdFbWk1enVl
OUJoeXkzMlV3Q2ZTQzhBCldHbUNpTEVwMkRDa2Q3T242Umk1d0NNPQo9NUhTVgotLS0tLUVORCBQ
R1AgU0lHTkFUVVJFLS0tLS0KCg==

--=__PartD1F660AD.0__=--

